File: src/Benchmark/Config/Env.hs
module Benchmark.Config.Env
  ( parseEnvFile
  , loadEnvVars
  , interpolateEnv
  )
where

import Control.Exception (IOException, try)
import Data.Bifunctor (bimap)
import Data.Map.Strict (Map)
import Data.Map.Strict qualified as Map
import Data.Maybe (mapMaybe)
import Data.Text (Text)
import Data.Text qualified as T
import Data.Text.IO qualified as TIO
import System.Directory (doesFileExist)
import System.Environment (getEnvironment)

{-| Parse a .env file into a map of key-value pairs.
Skips blank lines and lines starting with '#'.
Strips an optional 'export ' prefix; strips surrounding single or double quotes from values.
-}
parseEnvFile :: Text -> Map Text Text
parseEnvFile content = Map.fromList $ mapMaybe parseLine (T.lines content)
  where
    parseLine line =
      let stripped = T.strip line
       in if T.null stripped || T.isPrefixOf "#" stripped
            then Nothing
            else
              let noExport = maybe stripped T.stripStart (T.stripPrefix "export " stripped)
               in case T.breakOn "=" noExport of
                    (_, rest) | T.null rest -> Nothing
                    (key, valWithEq) ->
                      let key' = T.strip key
                          val' = stripQuotes (T.strip (T.drop 1 valWithEq))
                       in if T.null key' then Nothing else Just (key', val')
    stripQuotes t
      | T.length t >= 2 && T.head t == '"' && T.last t == '"' = T.init (T.tail t)
      | T.length t >= 2 && T.head t == '\'' && T.last t == '\'' = T.init (T.tail t)
      | otherwise = t

{-| Load environment variables from process env, .env, and .env.local.
Precedence (highest wins): .env.local > .env > process environment.
Missing files are silently ignored.
-}
loadEnvVars :: IO (Map Text Text)
loadEnvVars = do
  procEnv <- Map.fromList . map (bimap T.pack T.pack) <$> getEnvironment
  dotEnv <- readEnvFile ".env"
  dotEnvLocal <- readEnvFile ".env.local"
  return $ Map.unions [dotEnvLocal, dotEnv, procEnv]

readEnvFile :: FilePath -> IO (Map Text Text)
readEnvFile path = do
  exists <- doesFileExist path
  if exists
    then do
      result <- try (TIO.readFile path) :: IO (Either IOException Text)
      case result of
        Left _ -> return Map.empty
        Right content -> return (parseEnvFile content)
    else return Map.empty

{-| Interpolate \${VAR} patterns in the given text using the provided map.
Returns Left with an error message if a referenced variable is not defined.
An unclosed '\${' with no matching '}' is left as a literal.
-}
interpolateEnv :: Map Text Text -> Text -> Either String Text
interpolateEnv env = go mempty
  where
    go acc remaining =
      case T.breakOn "${" remaining of
        (before, rest)
          | T.null rest -> Right (acc <> before)
          | otherwise ->
              let afterOpen = T.drop 2 rest
               in case T.breakOn "}" afterOpen of
                    (_, noClose)
                      | T.null noClose ->
                          Right (acc <> before <> rest)
                    (varName, afterClose) ->
                      case Map.lookup varName env of
                        Nothing -> Left ("undefined environment variable: " <> T.unpack varName)
                        Just val -> go (acc <> before <> val) (T.drop 1 afterClose)

File: crates/gauntlet-core/src/config/env.rs
//! `.env` loading and `${VAR}` interpolation. See the crate docs for the
//! precedence rules and why this is hand-rolled.

use std::collections::HashMap;
use std::fs;
use std::path::Path;

/// Parse a `.env` file into key/value pairs, skipping blank and `#` lines and
/// stripping an optional `export ` prefix and matching surrounding quotes.
pub fn parse_env_file(content: &str) -> HashMap<String, String> {
    let mut map = HashMap::new();
    for line in content.lines() {
        let stripped = line.trim();
        if stripped.is_empty() || stripped.starts_with('#') {
            continue;
        }
        // Strip an optional `export ` prefix, then trim leading whitespace.
        let no_export = match stripped.strip_prefix("export ") {
            Some(rest) => rest.trim_start(),
            None => stripped,
        };
        // Split on the first '='. No '=' → skip the line.
        let Some(eq) = no_export.find('=') else {
            continue;
        };
        let key = no_export[..eq].trim();
        if key.is_empty() {
            continue;
        }
        let val = strip_quotes(no_export[eq + 1..].trim());
        map.insert(key.to_owned(), val.to_owned());
    }
    map
}

/// Strip a single pair of matching surrounding single or double quotes.
fn strip_quotes(t: &str) -> &str {
    let bytes = t.as_bytes();
    if bytes.len() >= 2 {
        let first = bytes[0];
        let last = bytes[bytes.len() - 1];
        if (first == b'"' && last == b'"') || (first == b'\'' && last == b'\'') {
            return &t[1..t.len() - 1];
        }
    }
    t
}

/// Merge process env, `.env`, and `.env.local`, highest precedence last. Missing
/// files are ignored.
pub fn load_env_vars() -> HashMap<String, String> {
    let mut merged: HashMap<String, String> = std::env::vars().collect();
    // `.env` overrides process env; `.env.local` overrides both.
    merged.extend(read_env_file(".env"));
    merged.extend(read_env_file(".env.local"));
    merged
}

fn read_env_file<P: AsRef<Path>>(path: P) -> HashMap<String, String> {
    match fs::read_to_string(path) {
        Ok(content) => parse_env_file(&content),
        Err(_) => HashMap::new(),
    }
}

/// Interpolate `${VAR}` patterns in `input`. An undefined reference returns
/// `Err(var_name)`, leaving the message to the caller.
pub fn interpolate_env(env: &HashMap<String, String>, input: &str) -> Result<String, String> {
    let mut out = String::with_capacity(input.len());
    let mut remaining = input;
    loop {
        match remaining.find("${") {
            None => {
                out.push_str(remaining);
                return Ok(out);
            }
            Some(open) => {
                let before = &remaining[..open];
                let after_open = &remaining[open + 2..];
                match after_open.find('}') {
                    // Unclosed `${` → emit the remainder verbatim, stop.
                    None => {
                        out.push_str(before);
                        out.push_str(&remaining[open..]);
                        return Ok(out);
                    }
                    Some(close) => {
                        let var_name = &after_open[..close];
                        match env.get(var_name) {
                            None => return Err(var_name.to_owned()),
                            Some(val) => {
                                out.push_str(before);
                                out.push_str(val);
                                remaining = &after_open[close + 1..];
                            }
                        }
                    }
                }
            }
        }
    }
}
